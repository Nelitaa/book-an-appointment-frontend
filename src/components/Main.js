import React, { useCallback, useEffect, useRef } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import '../styles/main.css';
import { fetchDoctors } from '../redux/doctors';
import { fetchReservations } from '../redux/reservations';
import SideBar from './SideBar';
import dots from '../assets/dots.png';
import left from '../assets/arrow-left.png';
import right from '../assets/arrow-right.png';

const Main = () => {
  const doctors = useSelector((state) => state.doctors);
  // const { length } = doctors;
  const dispatch = useDispatch();
  const doctorsContainerRef = useRef(null);

  // useEffect(() => {
  //   if (length === 0) {
  //     dispatch(fetchDoctors());
  //   }
  // }, [dispatch, length]);

  const retrieveData = useCallback(async () => {
    await dispatch(fetchReservations());
    await dispatch(fetchDoctors());
  }, [dispatch]);

  useEffect(() => {
    retrieveData();
  }, [retrieveData]);

  const scrollLeft = () => {
    if (doctorsContainerRef.current) {
      doctorsContainerRef.current.scrollLeft -= doctorsContainerRef.current.offsetWidth / 3;
    }
  };

  const scrollRight = () => {
    if (doctorsContainerRef.current) {
      doctorsContainerRef.current.scrollLeft += doctorsContainerRef.current.offsetWidth / 3;
    }
  };

  return (
    <div className="main-bar">
      <SideBar />
      <div className="main">
        <h1 className="doctor-list-title">BROWSE DOCTORS</h1>
        <p className="doctor-list-subtitle">Select a doctor to see details or reserve</p>
        <div className="dots-wrapper">
          <img src={dots} alt="dots-bar" className="dots-bar" />
        </div>
        {doctors.length === 0 ? ( // Add conditional rendering for empty doctors array
          <p className="no-doctors-message"><Link to="/add_doctor" className="link">Add a new doctor </Link></p> // Render the message
        ) : (
          <div className="doctors-container" ref={doctorsContainerRef}>
            {doctors.map((doctor) => (
              <div key={doctor.id} className="doctor-card">
                <img className="doctor-image" src={doctor.photo} alt="doctor" />
                <h2 className="doctor-name">
                  Dr.
                  {' '}
                  {doctor.name}
                </h2>
                <img src={dots} alt="dots-bar" className="dots-bar" />
                <p className="doctor-specialization">
                  {doctor.specialization}
                </p>
                <p className="doctor-city">
                  Based in
                  {' '}
                  {doctor.city}
                </p>
                <Link to={`/doctors/${doctor.id}`} className="doctor-details-link">
                  {/* Use Link to navigate to doctor details page */}
                  <button className="doctor-details" type="button">details</button>
                </Link>
              </div>
            ))}
          </div>
        )}
        <div className="arrow-buttons-container">
          <button className="arrow-button arrow-left" onClick={scrollLeft} type="button">
            <img src={left} alt="arrow-left" className="arrow-left" />
          </button>
          <button className="arrow-button arrow-right" onClick={scrollRight} type="button">
            <img src={right} alt="arrow-right" className="arrow-right" />
          </button>
        </div>
      </div>
    </div>
  );
};

export default Main;
