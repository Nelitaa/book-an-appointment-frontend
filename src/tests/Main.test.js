import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import store from '../redux/configureStore';
import Main from '../components/Main';

it('Main page should render correctly', () => {
  const result = render(
    <React.StrictMode>
      <Router>
        <Provider store={store}>
          <Main />
        </Provider>
      </Router>
    </React.StrictMode>,
  );
  expect(result).toMatchSnapshot();
});
