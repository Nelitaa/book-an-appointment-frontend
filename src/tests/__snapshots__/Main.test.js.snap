// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main page should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="main-bar"
      >
        <nav
          class="sidebar"
        >
          <div
            class="sidebar-header"
          >
            <div
              class="sidebar-header-icon"
            >
              <img
                alt="Our logo"
                class="logo"
                src="logo.png"
              />
            </div>
          </div>
          <ul
            class="sidebar-list"
          >
            <li
              class="sidebar-item"
            >
              <a
                class="link"
                href="/main"
              >
                DOCTORS
              </a>
            </li>
            <li
              class="sidebar-item"
            >
              <a
                class="link"
                href="/add_doctor"
              >
                ADD DOCTOR
              </a>
            </li>
            <li
              class="sidebar-item"
            >
              <a
                class="link"
                href="/doctors/delete"
              >
                DELETE DOCTOR
              </a>
            </li>
            <li
              class="sidebar-item"
            >
              <a
                class="link"
                href="/reserve"
              >
                RESERVE
              </a>
            </li>
            <li
              class="sidebar-item"
            >
              <a
                class="link"
                href="/reservations"
              >
                MY RESERVATIONS
              </a>
            </li>
          </ul>
          <a
            class="link sign-out-side"
            href="/signin"
          >
            Sign Out
          </a>
        </nav>
        <div
          class="main"
        >
          <h1
            class="doctor-list-title"
          >
            BROWSE DOCTORS
          </h1>
          <p
            class="doctor-list-subtitle"
          >
            Select a doctor to see details or reserve
          </p>
          <div
            class="dots-wrapper"
          >
            <img
              alt="dots-bar"
              class="dots-bar"
              src="dots.png"
            />
          </div>
          <p
            class="no-doctors-message"
          >
            <a
              class="link"
              href="/add_doctor"
            >
              Add a new doctor 
            </a>
          </p>
          <div
            class="arrow-buttons-container"
          >
            <button
              class="arrow-button arrow-left"
              type="button"
            >
              <img
                alt="arrow-left"
                class="arrow-left"
                src="arrow-left.png"
              />
            </button>
            <button
              class="arrow-button arrow-right"
              type="button"
            >
              <img
                alt="arrow-right"
                class="arrow-right"
                src="arrow-right.png"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="main-bar"
    >
      <nav
        class="sidebar"
      >
        <div
          class="sidebar-header"
        >
          <div
            class="sidebar-header-icon"
          >
            <img
              alt="Our logo"
              class="logo"
              src="logo.png"
            />
          </div>
        </div>
        <ul
          class="sidebar-list"
        >
          <li
            class="sidebar-item"
          >
            <a
              class="link"
              href="/main"
            >
              DOCTORS
            </a>
          </li>
          <li
            class="sidebar-item"
          >
            <a
              class="link"
              href="/add_doctor"
            >
              ADD DOCTOR
            </a>
          </li>
          <li
            class="sidebar-item"
          >
            <a
              class="link"
              href="/doctors/delete"
            >
              DELETE DOCTOR
            </a>
          </li>
          <li
            class="sidebar-item"
          >
            <a
              class="link"
              href="/reserve"
            >
              RESERVE
            </a>
          </li>
          <li
            class="sidebar-item"
          >
            <a
              class="link"
              href="/reservations"
            >
              MY RESERVATIONS
            </a>
          </li>
        </ul>
        <a
          class="link sign-out-side"
          href="/signin"
        >
          Sign Out
        </a>
      </nav>
      <div
        class="main"
      >
        <h1
          class="doctor-list-title"
        >
          BROWSE DOCTORS
        </h1>
        <p
          class="doctor-list-subtitle"
        >
          Select a doctor to see details or reserve
        </p>
        <div
          class="dots-wrapper"
        >
          <img
            alt="dots-bar"
            class="dots-bar"
            src="dots.png"
          />
        </div>
        <p
          class="no-doctors-message"
        >
          <a
            class="link"
            href="/add_doctor"
          >
            Add a new doctor 
          </a>
        </p>
        <div
          class="arrow-buttons-container"
        >
          <button
            class="arrow-button arrow-left"
            type="button"
          >
            <img
              alt="arrow-left"
              class="arrow-left"
              src="arrow-left.png"
            />
          </button>
          <button
            class="arrow-button arrow-right"
            type="button"
          >
            <img
              alt="arrow-right"
              class="arrow-right"
              src="arrow-right.png"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
